### PR TITLE
Outline the drain to heap logic in TinyVec::push

### PR DESCRIPTION
**Pull request by saethlin**
Per #39 

This is a strategy I've seen be very useful in other places; keeping code size down by forcibly outlining algorithmically cold paths tends to make a huge difference.

### CODE REVIEW

File: src/tinyvec.rs [hunk 1]
```diff
@@ -676,18 +676,25 @@ impl<A: Array> TinyVec<A> {
...
+  #[inline]
   pub fn push(&mut self, val: A::Item) {
-    let arr = match self {
-      TinyVec::Heap(v) => return v.push(val),
-      TinyVec::Inline(a) => a,
-    };
-
-    if let Some(x) = arr.try_push(val) {
+    #[inline(never)]
+    fn drain_to_heap_and_push<A: Array>(
```
Comment (Shnatsel) [inline src/tinyvec.rs:682]: You can also try `#[cold]` annotation, this should provide branch predictor hints for the match below.

Also, could you add a comment explaining why this code is in a separate function?

Comment (Lokathor) [reply src/tinyvec.rs:682]: A comment would be good about what's going on here.

Comment (saethlin) [reply src/tinyvec.rs:691]: I'm unable to measure a difference between `#[inline(never)]` and `#[cold]`. I've opted for `#[cold]` on the basis that if the optimizer somehow concludes that this function is profitable to inline, that is permitted.

**Review by Lokathor — APPROVED** [review]

**Comment by Lokathor** [comment]
Looks good. @saethlin i can merge this if you're satisfied.

**Comment by saethlin** [comment]
@Lokathor Yes please

**Pull request merged**